src/Source/AbstractPropertySourceIterator.php

```diff
@@ -0,0 +1,177 @@
+ */
+
+namespace Sonata\Exporter\Source;
+
+use Doctrine\ORM\Internal\Hydration\IterableResult;
+use Symfony\Component\PropertyAccess\Exception\UnexpectedTypeException;
+use Symfony\Component\PropertyAccess\PropertyAccess;
+use Symfony\Component\PropertyAccess\PropertyAccessor;
+use Symfony\Component\PropertyAccess\PropertyPath;
+
+abstract class AbstractPropertySourceIterator implements SourceIteratorInterface
+{
+    private const DATE_PARTS = [
+        'y' => 'Y',
+        'm' => 'M',
+        'd' => 'D',
+    ];
+    private const TIME_PARTS = [
+        'h' => 'H',
+        'i' => 'M',
+        's' => 'S',
+    ];
+
+    /**
+     * @var IterableResult
+     */
+    protected $iterator;
+
+    /**
+     * @var PropertyAccessor
+     */
+    protected $propertyAccessor;
+
+    /**
+     * @var string default DateTime format
+     */
+    protected $dateTimeFormat;
+
+    /**
+     * @var array<string>
+     */
+    protected $fields = [];
+
+    /**
+     * @param array<string> $fields Fields to export
+     */
+    public function __construct(array $fields, string $dateTimeFormat = 'r')
+    {
+        $this->propertyAccessor = PropertyAccess::createPropertyAccessor();
+        $this->fields = $fields;
+
+        $this->dateTimeFormat = $dateTimeFormat;
+    }
+
+    public function current()
+    {
+        $current = $this->iterator->current();
+
+        return $this->getCurrentData($current);
+    }
+
+    public function next(): void
+    {
+        $this->iterator->next();
+    }
+
+    public function key()
+    {
+        return $this->iterator->key();
+    }
+
+    public function valid(): bool
+    {
+        return $this->iterator->valid();
+    }
+
+    abstract public function rewind(): void;
+
+    public function setDateTimeFormat(string $dateTimeFormat): void
+    {
+        $this->dateTimeFormat = $dateTimeFormat;
+    }
+
+    public function getDateTimeFormat(): string
+    {
+        return $this->dateTimeFormat;
+    }
+
+    /**
+     * NEXT_MAJOR: Change the method visibility to private.
+     *
+     * @return string An ISO8601 duration
+     */
+    public function getDuration(\DateInterval $interval): string
+    {
+        $datePart = '';
+        foreach (self::DATE_PARTS as $datePartAttribute => $datePartAttributeString) {
+            if ($interval->$datePartAttribute !== 0) {
+                $datePart .= $interval->$datePartAttribute.$datePartAttributeString;
+            }
+        }
+
+        $timePart = '';
+        foreach (self::TIME_PARTS as $timePartAttribute => $timePartAttributeString) {
+            if ($interval->$timePartAttribute !== 0) {
+                $timePart .= $interval->$timePartAttribute.$timePartAttributeString;
+            }
+        }
+
+        if ('' === $datePart && '' === $timePart) {
+            return 'P0Y';
+        }
+
+        return 'P'.$datePart.('' !== $timePart ? 'T'.$timePart : '');
+    }
+
+    /**
+     * @param object|array $current
```

Comment:
array of what? `mixed`?

Reply:
I don't know, so yeah mixed maybe. I took the typehint from property accessor getValue.